### PR TITLE
rgw: support admin credentials in S3-related Keystone authentication.

### PR DESCRIPTION
**Pull request by rzarzynski**

### CODE REVIEW

**Comment by rzarzynski** [comment]
@yehudasa number of lines affected by this PR has been decreased several times. [Previous version](https://github.com/rzarzynski/ceph/tree/wip-admincreds-in-s3-keystone-old-1) is available as well.

**Comment by yehudasa** [comment]
@rzarzynski great, much better!

**Comment by zaitcev** [comment]
Looks good to me, but somewhat overcommented perhaps? There's a place where comment literally says "get auth token from Keystone" and the function name is get_keystone_admin_token().

**Comment by rzarzynski** [comment]
@yehudasa rebased to master and improved spacing according to Matt's comment.

Today I've tested the patch using s3-tests, Tempest and manually against radosgw in many different configurations (Keystone/admin token, Keystone/admin credentials, internal RADOS-based auth). Tests showed there is no regression here. IMO the PR is ready to merge.

File: src/rgw/rgw_rest_s3.cc [hunk 1]
```diff
@@ -2560,8 +2564,16 @@ int RGW_Auth_S3_Keystone_ValidateToken::validate_s3token(const string& auth_id,
     keystone_url.append("/");
   keystone_url.append("v2.0/s3tokens");
 
+  /* get authentication token for Keystone. */
+  string admin_token_id;
+  int r = RGWSwift::get_keystone_admin_token(cct, admin_token_id);
+  if (r < 0) {
+    dout(2) << "s3 keystone: cannot get token for keystone access" << dendl;
```
Comment (yehudasa) [inline src/rgw/rgw_rest_s3.cc]: @rzarzynski one small nitpick, please change this dout(2) to ldout(cct, 2). Once that's done I'll merge it.

**Comment by rzarzynski** [comment]
@yehudasa thanks for pointing the dout. Pushed amended commit.

**Pull request merged**